File: data-structures/5_trees/1_binary_search_trees/bst.ts
class BSTNode {
  value: number;
  left: BSTNode | null;
  right: BSTNode | null;
  parent: BSTNode | null;

  constructor(value: number) {
    this.value = value;
    this.left = null;
    this.right = null;
    this.parent = null;
  }

  addNode(value: number): BSTNode | undefined {
    if (value < this.value) {
      if (this.left) {
        return this.left.addNode(value);
      }
      const newNode = new BSTNode(value);
      this.left = newNode;
      newNode.parent = this;
    } else {
      if (this.right) {
        return this.right.addNode(value);
      }
      const newNode = new BSTNode(value);
      this.right = newNode;
      newNode.parent = this;
    }
  }
  // Left → Root → Right
  inOrderTraversal(node: BSTNode = this) {
    node.left && this.inOrderTraversal(node.left);
    console.log(node.value);
    node.right && this.inOrderTraversal(node.right);
  }

  // Root → Left → Right
  preOrderTraversal(node: BSTNode = this) {
    console.log(node.value);
    node.left && this.preOrderTraversal(node.left);
    node.right && this.preOrderTraversal(node.right);
  }

  // Left → Right → Root
  postOrderTraversal(node: BSTNode = this) {
    node.left && this.postOrderTraversal(node.left);
    node.right && this.postOrderTraversal(node.right);
    console.log(node.value);
  }

  // BFS
  BFSTraversal() {
    let result: Array<number> = [];
    let queue = [this as BSTNode];
    while (queue.length > 0) {
      let targetNode = queue.shift();
      if (targetNode) {
        result.push(targetNode.value);
        targetNode.left && queue.push(targetNode.left);
        targetNode.right && queue.push(targetNode.right);
      }
    }
    return result;
  }
}

const node = new BSTNode(10);

node.addNode(5);
node.addNode(14);
node.addNode(2);
node.addNode(6);
node.addNode(24);

// node.inOrderTraversal();
// node.preOrderTraversal();
// node.postOrderTraversal();
console.log(node.BFSTraversal());

File: data-structures/5_trees/1_binary_search_trees/bst.py
# class node:
#     def __init__(self, value):
#        self.value = value
#        self.left = None
#        self.right = None
#        self.parent = None

class BST:
    def __init__(self, value):
       self.value = value
       self.left = None
       self.right = None
       self.parent = None
    
    def addNode(self, value):
        if(value < self.value):
            if(self.left):
                return self.left.addNode(value)
            newNode = BST(value)
            self.left = newNode
            newNode.parent = self
        else:
            if(self.right):
                return self.right.addNode(value)
            newNode = BST(value)
            self.right = newNode
            newNode.parent = self
    
    # Left → Root → Right
    def inOrderTraversal(self):
        if(self.left):
            self.left.inOrderTraversal()
        print(self.value)
        if(self.right):
            self.right.inOrderTraversal()

    # Root → Left → Right
    def preOrderTraversal(self):
        print(self.value)
        if(self.left):
            self.left.inOrderTraversal()
        if(self.right):
            self.right.inOrderTraversal()

    # Left → Right → Root
    def postOrderTraversal(self):
        if(self.left):
            self.left.inOrderTraversal()
        if(self.right):
            self.right.inOrderTraversal()
        print(self.value)
    
    def BFSTraversal(self):
        result = []
        queue = [self]
        while(len(queue)):
            targetNode = queue.pop(0)
            result.append(targetNode.value)
            if(targetNode.left):
                queue.append(targetNode.left)
            if(targetNode.right):
                queue.append(targetNode.right)
        return result




node = BST(10)

node.addNode(5)
node.addNode(14)
node.addNode(2)
node.addNode(6)
node.addNode(24)

# node.inOrderTraversal()
# node.preOrderTraversal()
# node.postOrderTraversal()

print(node.BFSTraversal())


